packages/udt/src/udt/index.ts

```diff
@@ -1,306 +1,394 @@
-import { ccc } from "@ckb-ccc/core";
-import { ssri } from "@ckb-ccc/ssri";
-
-/**
- * Represents a User Defined Token (UDT) script compliant with the SSRI protocol.
- *
- * This class provides a comprehensive implementation for interacting with User Defined Tokens,
- * supporting various token operations such as querying metadata, checking balances, and performing transfers.
- * It supports both SSRI-compliant UDTs and legacy sUDT/xUDT standard tokens.
- *
- * @public
- * @category Blockchain
- * @category Token
- */
-export class Udt extends ssri.Trait {
-  public readonly script: ccc.Script;
-
-  /**
-   * Constructs a new UDT (User Defined Token) script instance.
-   * By default it is a SSRI-compliant UDT. By providing `xudtType`, it is compatible with the legacy xUDT.
-   *
-   * @param executor - The SSRI executor instance.
-   * @param code - The script code cell of the UDT.
-   * @param script - The type script of the UDT.
-   * @example
-   * ```typescript
-   * const udt = new Udt(executor, code, script);
-   * ```
-   */
+import { ccc, mol } from "@ckb-ccc/core";
+
+export class ErrorUdtInsufficientCoin extends Error {
   constructor(
-    code: ccc.OutPointLike,
-    script: ccc.ScriptLike,
-    config?: {
-      executor?: ssri.Executor | null;
-    } | null,
+    public readonly amount: ccc.FixedPoint,
+    public readonly udtManager: UdtManager,
   ) {
-    super(code, config?.executor);
-    this.script = ccc.Script.from(script);
+    super(
+      `Insufficient coin, need ${ccc.fixedPointToString(amount, udtManager.decimals)} extra ${udtManager.symbol} coin`,
+    );
   }
+}
 
-  /**
-   * Retrieves the human-readable name of the User Defined Token.
-   *
-   * @returns A promise resolving to the token's name.
-   */
-  async name(
-    context?: ssri.ContextScript,
-  ): Promise<ssri.ExecutorResponse<string | undefined>> {
-    if (this.executor) {
-      const res = await this.executor.runScriptTry(this.code, "UDT.name", [], {
-        script: this.script,
-        ...context,
-      });
-      if (res) {
-        return res.map((res) => ccc.bytesTo(res, "utf8"));
-      }
-    }
+export class UdtManager {
+  constructor(
+    public readonly script: ccc.Script,
+    public readonly cellDeps: ccc.CellDep[],
+    public readonly dataLenRange: [number, number],
+    public readonly name: string,
+    public readonly symbol: string,
+    public readonly decimals: number,
+    public readonly icon: string,
+  ) {}
 
-    return ssri.ExecutorResponse.new(undefined);
-  }
+  async *find(
+    client: ccc.Client,
+    locks: ccc.Script[],
```

Comment:
I'll always prefer `Signer`. While it might not be the best name for all cases (Like `Watcher` or sth might be better than `SignerReadonly`), it does reduce the aspects that newcomers have to learn. Besides, the utilities implemented in `Signer` also help us to avoid repeating ourselves.

We should use something like `SignerCkbScriptListReadonly` instead of `Script[]` everywhere, for special cases.

Reply:
We can have them.